File: clients/sdk/src/metadata.ts
/**
 * Payment metadata: the ISO 20022 detail that no longer goes on the chain.
 *
 * `purpose_code` and `remittance_information` are free-text fields, and in
 * practice they are where an operator puts a customer's name. Written to this
 * chain they were public, permanent and unerasable — the exposure under
 * Nigeria's NDPA, Ghana's DPA, POPIA and the GDPR. They now travel as a payload
 * held here, and the chain records only a hash of it.
 *
 * The hash is what makes the payload worth anything. Without it the payload is
 * a claim by whoever happens to be storing it, and a party in a dispute could
 * produce whichever version of the remittance line suited them. With it, a
 * payload either hashes to what the block says or it is not the payload.
 *
 * Two things this deliberately does not do, stated here because assuming
 * otherwise is how a pilot goes wrong:
 *
 * - **It does not encrypt.** A hash proves integrity and reveals nothing, but
 *   it also lets nobody read. The viewing keys that let the payer, the payee
 *   and the regulator decrypt are the next workstream; see
 *   docs/scope/confidentiality.md.
 * - **It does not deliver.** This store is `localStorage`, so it is one browser
 *   on one origin. The payer's device holds the payload the payer wrote; the
 *   payee's device does not. Until a shared encrypted store exists, the payload
 *   has to reach the other side by whatever channel already carries the
 *   remittance advice, and only the hash makes it checkable when it arrives.
 */

import { PaymentMetadata } from './generated/blockchain/paymsg/v1/payment_metadata.ts';

/** Length of the salt, matching MetadataSaltLength in x/paymsg/types. */
export const METADATA_SALT_BYTES = 32;

/** Length of the recorded hash, matching MetadataHashLength in x/paymsg/types. */
export const METADATA_HASH_BYTES = 32;

/** ISO 20022 limits, the same ones the chain applies to the plaintext fields. */
const MAX_PURPOSE_CODE = 4;
const MAX_REMITTANCE = 140;

export interface PaymentMetadataPayload {
  /** 32 random bytes, fresh for every payment. */
  salt: Uint8Array;
  purposeCode: string;
  remittanceInformation: string;
}

/**
 * Build a payload with a fresh salt.
 *
 * The salt is not decoration. A purpose code is four characters from a
 * published list, so the hash of an unsalted one is not a fingerprint but a
 * lookup table anybody can build in a second — and the ledger is public and
 * permanent, so they have unlimited time to try. It is regenerated per payment
 * rather than per account because a reused salt makes two payments carrying the
 * same detail hash identically, which tells an observer they match without ever
 * revealing what they say.
 */
export function newPaymentMetadata(
  purposeCode: string,
  remittanceInformation: string,
): PaymentMetadataPayload {
  if (purposeCode.length > MAX_PURPOSE_CODE) {
    throw new Error(`purposeCode must be at most ${MAX_PURPOSE_CODE} characters`);
  }
  if (remittanceInformation.length > MAX_REMITTANCE) {
    throw new Error(`remittanceInformation must be at most ${MAX_REMITTANCE} characters`);
  }

  const salt = new Uint8Array(METADATA_SALT_BYTES);
  crypto.getRandomValues(salt);
  return { salt, purposeCode, remittanceInformation };
}

/**
 * Hash a payload into the value that goes on-chain.
 *
 * The bytes hashed are the generated protobuf encoding, not a string this file
 * assembles. That is the only reason the chain and this package agree: a
 * hand-written serialiser here and a protobuf one in the keeper would produce
 * the same answer right up until somebody added a field, and a hash that
 * disagrees proves nothing while looking exactly like a hash that works.
 */
export async function metadataHash(payload: PaymentMetadataPayload): Promise<Uint8Array> {
  if (payload.salt.length !== METADATA_SALT_BYTES) {
    throw new Error(
      `salt must be ${METADATA_SALT_BYTES} bytes, got ${payload.salt.length}; an unsalted hash publishes the field it was meant to hide`,
    );
  }

  const encoded = PaymentMetadata.encode({
    salt: payload.salt,
    purposeCode: payload.purposeCode,
    remittanceInformation: payload.remittanceInformation,
  }).finish();

  // Copied into a standalone buffer rather than handed over as-is. protobufjs
  // writes into a pooled backing array, so the encoder's view can sit inside a
  // larger buffer shared with whatever was encoded before it — digesting that
  // would hash the neighbours too, and only sometimes.
  const bytes = new Uint8Array(encoded.length);
  bytes.set(encoded);

  const digest = await crypto.subtle.digest('SHA-256', bytes.buffer);
  return new Uint8Array(digest);
}

/**
 * Check a payload against the hash a payment recorded.
 *
 * Anyone who receives a payload from anywhere — the other participant, a
 * backup, this store after a browser upgrade — should run this before believing
 * it. The hash comes from the block; the payload does not.
 */
export async function verifyMetadata(
  payload: PaymentMetadataPayload,
  recorded: Uint8Array,
): Promise<boolean> {
  if (recorded.length !== METADATA_HASH_BYTES) return false;

  const computed = await metadataHash(payload);
  let differing = 0;
  for (let i = 0; i < METADATA_HASH_BYTES; i++) differing |= computed[i]! ^ recorded[i]!;
  return differing === 0;
}

// ------------------------------------------------------------------- storage

const STORAGE_KEY = 'yamale.paymsg.metadata.v1';

/**
 * Payloads are filed under the chain's own key for a payment — (instructing
 * participant, end-to-end id) — rather than a local id, so a record fetched
 * from the node leads straight to its payload without a second index to keep
 * in step.
 */
function storageKey(instructingParticipant: string, endToEndId: string): string {
  return `${instructingParticipant}/${endToEndId}`;
}

/** The salt is bytes and localStorage holds strings, so it is stored base64. */
interface StoredPayload {
  salt: string;
  purposeCode: string;
  remittanceInformation: string;
  savedAt: string;
}

type Store = Record<string, StoredPayload>;

function read(): Store {
  try {
    const raw = localStorage.getItem(STORAGE_KEY);
    return raw ? (JSON.parse(raw) as Store) : {};
  } catch {
    // A corrupt store must not take the page down. Returning empty loses the
    // detail, not the payment: the chain still holds the hash, so a payload
    // recovered from anywhere else can still be proven to be the right one.
    return {};
  }
}

function write(store: Store): void {
  localStorage.setItem(STORAGE_KEY, JSON.stringify(store));
  window.dispatchEvent(new CustomEvent('yamale:paymsg-metadata'));
}

function toBase64(bytes: Uint8Array): string {
  let binary = '';
  for (const byte of bytes) binary += String.fromCharCode(byte);
  return btoa(binary);
}

function fromBase64(encoded: string): Uint8Array {
  const binary = atob(encoded);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
  return bytes;
}

/** File a payload against the payment that recorded its hash. */
export function savePaymentMetadata(
  instructingParticipant: string,
  endToEndId: string,
  payload: PaymentMetadataPayload,
): void {
  const store = read();
  store[storageKey(instructingParticipant, endToEndId)] = {
    salt: toBase64(payload.salt),
    purposeCode: payload.purposeCode,
    remittanceInformation: payload.remittanceInformation,
    savedAt: new Date().toISOString(),
  };
  write(store);
}

/** Retrieve a payload. Verify it against the chain's hash before trusting it. */
export function loadPaymentMetadata(
  instructingParticipant: string,
  endToEndId: string,
): PaymentMetadataPayload | undefined {
  const stored = read()[storageKey(instructingParticipant, endToEndId)];
  if (!stored) return undefined;

  return {
    salt: fromBase64(stored.salt),
    purposeCode: stored.purposeCode,
    remittanceInformation: stored.remittanceInformation,
  };
}

File: x/paymsg/types/metadata.go
package types

import (
	"bytes"
	"crypto/rand"
	"crypto/sha256"

	errorsmod "cosmossdk.io/errors"
)

const (
	// MetadataHashLength is the length of the commitment the chain records.
	//
	// Checked rather than assumed. A hash of any other length is not a hash —
	// it is whatever the sender felt like putting in the field, stored forever
	// in a statement record where it will read as a commitment to somebody
	// reconciling against it years later.
	MetadataHashLength = sha256.Size

	// MetadataSaltLength is the salt every payload carries.
	//
	// 32 bytes rather than 8 or 16 because the salt is the only thing standing
	// between a hash of a four-character purpose code and a lookup table, and
	// the ledger is public and permanent: an attacker gets unlimited time and
	// unlimited guesses, so the margin has to be one nobody can grind through
	// rather than one that is merely inconvenient today.
	MetadataSaltLength = 32
)

// NewPaymentMetadata builds a payload with a fresh salt.
//
// It belongs to whoever is composing the payment — a wallet, a participant's
// back office — and never to the chain. A node calling this would be
// generating a secret it then has to keep, which is the arrangement the whole
// design exists to avoid; the keeper only ever sees the hash.
func NewPaymentMetadata(purposeCode, remittanceInformation string) (PaymentMetadata, error) {
	salt := make([]byte, MetadataSaltLength)
	if _, err := rand.Read(salt); err != nil {
		return PaymentMetadata{}, errorsmod.Wrap(err, "unable to generate a metadata salt")
	}

	return PaymentMetadata{
		Salt:                  salt,
		PurposeCode:           purposeCode,
		RemittanceInformation: remittanceInformation,
	}, nil
}

// Hash returns the value that goes on-chain in MsgSendPayment.metadata_hash.
//
// The salt is required here rather than defaulted, because a payload that
// reached this function without one has already lost the property the hash was
// added for, and returning a usable hash anyway would hide that until the
// ledger was public.
func (m PaymentMetadata) Hash() ([]byte, error) {
	if len(m.Salt) != MetadataSaltLength {
		return nil, errorsmod.Wrapf(ErrInvalidMetadata,
			"salt must be %d bytes, got %d", MetadataSaltLength, len(m.Salt))
	}
	if err := ValidateMetadataFields(m.PurposeCode, m.RemittanceInformation); err != nil {
		return nil, err
	}

	bz, err := m.Marshal()
	if err != nil {
		return nil, err
	}
	sum := sha256.Sum256(bz)
	return sum[:], nil
}

// VerifyMetadata reports whether a payload held off-chain is the one a payment
// recorded.
//
// This is the only check that gives the off-chain store any authority. Without
// it the payload is a claim by whoever is storing it, and a participant in a
// dispute could produce whichever version of the remittance information suited
// them; with it, either the payload hashes to what the block says or it is not
// the payload.
func VerifyMetadata(payload PaymentMetadata, recorded []byte) error {
	if len(recorded) != MetadataHashLength {
		return errorsmod.Wrapf(ErrInvalidMetadata,
			"recorded hash must be %d bytes, got %d", MetadataHashLength, len(recorded))
	}

	computed, err := payload.Hash()
	if err != nil {
		return err
	}
	if !bytes.Equal(computed, recorded) {
		return errorsmod.Wrap(ErrInvalidMetadata,
			"payload does not hash to the value recorded on-chain, so it is not the payload this payment carried")
	}
	return nil
}

// ValidateMetadataFields applies the ISO 20022 limits to a payload.
//
// The same limits as the plaintext fields they replace. Moving the detail
// off-chain does not make it unbounded: a payload that exceeds what the
// receiving system can hold still cannot be delivered, and the limits are
// worth failing on while the sender is still there to be told.
func ValidateMetadataFields(purposeCode, remittanceInformation string) error {
	if len(purposeCode) > MaxPurposeCodeLength {
		return errorsmod.Wrapf(ErrInvalidMetadata,
			"purpose_code must be at most %d characters, got %d", MaxPurposeCodeLength, len(purposeCode))
	}
	if len(remittanceInformation) > MaxRemittanceLength {
		return errorsmod.Wrapf(ErrInvalidMetadata,
			"remittance_information must be at most %d characters, got %d", MaxRemittanceLength, len(remittanceInformation))
	}
	return nil
}
